File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/chat/components/ToolCallViews.kt
package com.runanywhere.runanywhereai.presentation.chat.components

import androidx.compose.animation.core.*
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import com.runanywhere.runanywhereai.domain.models.ToolCallInfo
import com.runanywhere.runanywhereai.ui.theme.AppColors
import com.runanywhere.runanywhereai.ui.theme.AppTypography

/**
 * Tool call indicator button - matches iOS ToolCallIndicator
 * Shows a small chip with the tool name that can be tapped to see details
 */
@Composable
fun ToolCallIndicator(
    toolCallInfo: ToolCallInfo,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val backgroundColor = if (toolCallInfo.success) {
        AppColors.primaryAccent.copy(alpha = 0.1f)
    } else {
        AppColors.primaryOrange.copy(alpha = 0.1f)
    }
    
    val borderColor = if (toolCallInfo.success) {
        AppColors.primaryAccent.copy(alpha = 0.3f)
    } else {
        AppColors.primaryOrange.copy(alpha = 0.3f)
    }
    
    val iconTint = if (toolCallInfo.success) {
        AppColors.primaryAccent
    } else {
        AppColors.primaryOrange
    }

    Surface(
        modifier = modifier
            .clip(RoundedCornerShape(8.dp))
            .clickable { onTap() },
        color = backgroundColor,
        shape = RoundedCornerShape(8.dp),
    ) {
        Row(
            modifier = Modifier
                .border(0.5.dp, borderColor, RoundedCornerShape(8.dp))
                .padding(horizontal = 10.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Icon(
                imageVector = if (toolCallInfo.success) Icons.Default.Build else Icons.Default.Warning,
                contentDescription = null,
                modifier = Modifier.size(12.dp),
                tint = iconTint,
            )
            Text(
                text = toolCallInfo.toolName,
                style = AppTypography.caption2,
                color = AppColors.textSecondary,
                maxLines = 1,
            )
        }
    }
}

/**
 * Tool call detail sheet - matches iOS ToolCallDetailSheet
 * Shows full details of a tool call including arguments and results as JSON
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ToolCallDetailSheet(
    toolCallInfo: ToolCallInfo,
    onDismiss: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = MaterialTheme.colorScheme.surface,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp)
                .padding(bottom = 32.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(20.dp),
        ) {
            // Header
            Text(
                text = "Tool Call",
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.onSurface,
            )
            
            // Status section
            StatusSection(success = toolCallInfo.success)
            
            // Tool name section
            DetailSection(title = "Tool", content = toolCallInfo.toolName)
            
            // Arguments section
            CodeSection(title = "Arguments", code = toolCallInfo.arguments)
            
            // Result section (if available)
            toolCallInfo.result?.let { result ->
                CodeSection(title = "Result", code = result)
            }
            
            // Error section (if available)
            toolCallInfo.error?.let { error ->
                DetailSection(title = "Error", content = error, isError = true)
            }
            
            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}

@Composable
private fun StatusSection(success: Boolean) {
    val backgroundColor = if (success) {
        AppColors.statusGreen.copy(alpha = 0.1f)
    } else {
        AppColors.primaryRed.copy(alpha = 0.1f)
    }
    
    val iconTint = if (success) AppColors.statusGreen else AppColors.primaryRed
    
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(backgroundColor, RoundedCornerShape(12.dp))
            .padding(16.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Icon(
            imageVector = if (success) Icons.Default.CheckCircle else Icons.Default.Cancel,
            contentDescription = null,
            modifier = Modifier.size(24.dp),
            tint = iconTint,
        )
        Text(
            text = if (success) "Success" else "Failed",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

@Composable
private fun DetailSection(
    title: String,
    content: String,
    isError: Boolean = false,
) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(
            text = title,
            style = AppTypography.caption,
            color = AppColors.textSecondary,
        )
        Text(
            text = content,
            style = MaterialTheme.typography.bodyMedium,
            color = if (isError) AppColors.primaryRed else MaterialTheme.colorScheme.onSurface,
        )
    }
}

@Composable
private fun CodeSection(
    title: String,
    code: String,
) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(
            text = title,
            style = AppTypography.caption,
            color = AppColors.textSecondary,
        )
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(
                    MaterialTheme.colorScheme.surfaceVariant,
                    RoundedCornerShape(8.dp)
                )
                .padding(12.dp)
                .horizontalScroll(rememberScrollState())
        ) {
            Text(
                text = code,
                style = MaterialTheme.typography.bodyMedium.copy(fontFamily = FontFamily.Monospace),
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

/**
 * Tool calling active indicator - matches iOS ToolCallingActiveIndicator
 * Shows animated indicator when tool is being called
 */
@Composable
fun ToolCallingActiveIndicator(
    modifier: Modifier = Modifier,
) {
    val infiniteTransition = rememberInfiniteTransition(label = "rotation")
    val rotation by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(2000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "rotation"
    )
    
    Row(
        modifier = modifier
            .background(
                AppColors.primaryAccent.copy(alpha = 0.1f),
                RoundedCornerShape(8.dp)
            )
            .padding(horizontal = 10.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Icon(
            imageVector = Icons.Default.Settings,
            contentDescription = null,
            modifier = Modifier
                .size(12.dp)
                .graphicsLayer { rotationZ = rotation },
            tint = AppColors.primaryAccent,
        )
        Text(
            text = "Calling tool...",
            style = AppTypography.caption2,
            color = AppColors.textSecondary,
        )
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/chat/tool_call_views.dart
import 'package:flutter/material.dart';
import 'package:runanywhere_ai/core/design_system/app_colors.dart';
import 'package:runanywhere_ai/core/design_system/app_spacing.dart';
import 'package:runanywhere_ai/core/design_system/typography.dart';

/// Tool Call Info model (mirroring iOS ToolCallInfo)
class ToolCallInfo {
  final String toolName;
  final String arguments; // JSON string for display
  final String? result; // JSON string for display
  final bool success;
  final String? error;

  const ToolCallInfo({
    required this.toolName,
    required this.arguments,
    this.result,
    required this.success,
    this.error,
  });
}

/// Tool Call Indicator (mirroring iOS ToolCallIndicator)
///
/// A tappable badge that shows tool call status and opens detail sheet.
class ToolCallIndicator extends StatelessWidget {
  final ToolCallInfo toolCallInfo;
  final VoidCallback onTap;

  const ToolCallIndicator({
    super.key,
    required this.toolCallInfo,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final isSuccess = toolCallInfo.success;
    final accentColor =
        isSuccess ? AppColors.primaryAccent : AppColors.primaryOrange;

    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(
          horizontal: 10,
          vertical: 6,
        ),
        decoration: BoxDecoration(
          color: accentColor.withOpacity(0.1),
          borderRadius: BorderRadius.circular(8),
          border: Border.all(
            color: accentColor.withOpacity(0.3),
            width: 0.5,
          ),
        ),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(
              isSuccess ? Icons.build_outlined : Icons.warning_amber,
              size: 12,
              color: accentColor,
            ),
            const SizedBox(width: 6),
            Text(
              toolCallInfo.toolName,
              style: AppTypography.caption2(context).copyWith(
                color: AppColors.textSecondary(context),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

/// Tool Call Detail Sheet (mirroring iOS ToolCallDetailSheet)
///
/// Shows full details of a tool call.
class ToolCallDetailSheet extends StatelessWidget {
  final ToolCallInfo toolCallInfo;

  const ToolCallDetailSheet({
    super.key,
    required this.toolCallInfo,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
        color: AppColors.backgroundPrimary(context),
        borderRadius: const BorderRadius.vertical(
          top: Radius.circular(20),
        ),
      ),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          // Handle bar
          Container(
            width: 40,
            height: 4,
            margin: const EdgeInsets.only(top: 12),
            decoration: BoxDecoration(
              color: AppColors.separator(context),
              borderRadius: BorderRadius.circular(2),
            ),
          ),

          // Header
          Padding(
            padding: const EdgeInsets.all(AppSpacing.large),
            child: Row(
              children: [
                Text(
                  'Tool Call',
                  style: AppTypography.headline(context),
                ),
                const Spacer(),
                TextButton(
                  onPressed: () => Navigator.pop(context),
                  child: const Text('Done'),
                ),
              ],
            ),
          ),

          const Divider(height: 1),

          // Content
          Flexible(
            child: SingleChildScrollView(
              padding: const EdgeInsets.all(AppSpacing.large),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  // Status section
                  _buildStatusSection(context),

                  const SizedBox(height: AppSpacing.xLarge),

                  // Tool name
                  _buildDetailSection(
                    context,
                    title: 'Tool',
                    content: toolCallInfo.toolName,
                  ),

                  const SizedBox(height: AppSpacing.xLarge),

                  // Arguments
                  _buildCodeSection(
                    context,
                    title: 'Arguments',
                    code: toolCallInfo.arguments,
                  ),

                  // Result (if present)
                  if (toolCallInfo.result != null) ...[
                    const SizedBox(height: AppSpacing.xLarge),
                    _buildCodeSection(
                      context,
                      title: 'Result',
                      code: toolCallInfo.result!,
                    ),
                  ],

                  // Error (if present)
                  if (toolCallInfo.error != null) ...[
                    const SizedBox(height: AppSpacing.xLarge),
                    _buildDetailSection(
                      context,
                      title: 'Error',
                      content: toolCallInfo.error!,
                      isError: true,
                    ),
                  ],

                  const SizedBox(height: AppSpacing.xxLarge),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildStatusSection(BuildContext context) {
    final isSuccess = toolCallInfo.success;
    final statusColor =
        isSuccess ? AppColors.statusGreen : AppColors.primaryRed;

    return Container(
      padding: const EdgeInsets.all(AppSpacing.large),
      decoration: BoxDecoration(
        color: statusColor.withOpacity(0.1),
        borderRadius: BorderRadius.circular(12),
      ),
      child: Row(
        children: [
          Icon(
            isSuccess ? Icons.check_circle : Icons.cancel,
            size: 24,
            color: statusColor,
          ),
          const SizedBox(width: 10),
          Text(
            isSuccess ? 'Success' : 'Failed',
            style: AppTypography.headline(context),
          ),
        ],
      ),
    );
  }

  Widget _buildDetailSection(
    BuildContext context, {
    required String title,
    required String content,
    bool isError = false,
  }) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          title,
          style: AppTypography.caption(context).copyWith(
            color: AppColors.textSecondary(context),
          ),
        ),
        const SizedBox(height: 8),
        Text(
          content,
          style: AppTypography.body(context).copyWith(
            color: isError ? AppColors.primaryRed : null,
          ),
        ),
      ],
    );
  }

  Widget _buildCodeSection(
    BuildContext context, {
    required String title,
    required String code,
  }) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          title,
          style: AppTypography.caption(context).copyWith(
            color: AppColors.textSecondary(context),
          ),
        ),
        const SizedBox(height: 8),
        Container(
          width: double.infinity,
          padding: const EdgeInsets.all(12),
          decoration: BoxDecoration(
            color: AppColors.backgroundSecondary(context),
            borderRadius: BorderRadius.circular(8),
          ),
          child: Text(
            code,
            style: AppTypography.monospaced,
          ),
        ),
      ],
    );
  }
}

/// Tool Calling Active Indicator (mirroring iOS ToolCallingActiveIndicator)
///
/// Shows "Calling tool..." with animated gear icon.
class ToolCallingActiveIndicator extends StatefulWidget {
  const ToolCallingActiveIndicator({super.key});

  @override
  State<ToolCallingActiveIndicator> createState() =>
      _ToolCallingActiveIndicatorState();
}

class _ToolCallingActiveIndicatorState extends State<ToolCallingActiveIndicator>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 2),
      vsync: this,
    )..repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(
        horizontal: 10,
        vertical: 6,
      ),
      decoration: BoxDecoration(
        color: AppColors.primaryAccent.withOpacity(0.1),
        borderRadius: BorderRadius.circular(8),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          RotationTransition(
            turns: _controller,
            child: const Icon(
              Icons.settings,
              size: 12,
              color: AppColors.primaryBlue,
            ),
          ),
          const SizedBox(width: 6),
          Text(
            'Calling tool...',
            style: AppTypography.caption2(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
          ),
        ],
      ),
    );
  }
}

/// Tool Calling Badge (mirroring iOS toolCallingBadge)
///
/// Shows "Tools enabled" badge above chat input.
class ToolCallingBadge extends StatelessWidget {
  final int toolCount;

  const ToolCallingBadge({
    super.key,
    required this.toolCount,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(
        horizontal: 10,
        vertical: 4,
      ),
      decoration: BoxDecoration(
        color: AppColors.primaryAccent.withOpacity(0.1),
        borderRadius: BorderRadius.circular(6),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          const Icon(
            Icons.build_outlined,
            size: 10,
            color: AppColors.primaryBlue,
          ),
          const SizedBox(width: 6),
          Text(
            'Tools enabled ($toolCount)',
            style: AppTypography.caption2(context).copyWith(
              color: AppColors.primaryAccent,
            ),
          ),
        ],
      ),
    );
  }
}
